CMake: update docs/application/application.rst

mbolivar opened the pull request:
This series re-works the application development primer for CMake compatibility, among other general improvements.

Commit:
doc: getting_started: add ref target for running hello world

This makes it possible to point users at a canonical location for how
to use zephyr-env.sh, etc.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: CMake updates for overview and creation docs

Update the initial application overview and the basic "how to create
an application" sections for the CMake transition. This is worth doing
on its own, and also enables other fixes and improvements to below
sections.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: touch up cmake overview

Fix up some grammar issues and add more reST formatting.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: re-work CMakeLists.txt / configuration division

Fix a few issues related to the description of an application's
CMakeLists.txt file, and how that relates to its configuraiton.

Make sure the section "Application CMakeLists.txt" appears under the
parent "CMake" section, instead of on its own. The order in which
lines appear in the application CMakeLists.txt is important, and
that's not coming through clearly, so try to improve that. Document
how the values for BOARD and CONF_FILE are determined by
boilerplate.cmake here. Also document usage of KCONFIG_ROOT, as its
Kbuild-based equivalent is something that users ask about.

Merge some content from the following section "Application
Configuration" into the appropriate places, to keep the document flow
working. Add references in "Application Configuration" to definitions
provided previously in the document, for clarity.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: improve Kconfig setting documentation

Now that the above sections more clearly define the behavior of
CONF_FILE, the section on writing a .conf can be cleaned up and made
into a simple primer on syntax.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: improvements to "Overriding Default Configuration"

Make this a bit less Unix-centric, changing the CMake invocation lines
to only use documented parameters that continue to work across all
CMake versions (-B doesn't work everywhere, and -H means "help" in
recent versions of CMake).

Handle some 80-column cleanliness. Add documentation about the search
key, /. Be a bit more explicit about the steps.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: trivial fixups to "Rebuild an Application"

Make clear that a terminal is being used, and use a Sphinx :ref:
instead of a reST link.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: move build/run/debug docs after overview

This document's flow could be improved a bit. The overview section is
followed by somewhat dense reference material, which is then followed
by step-by-step instructions central to the application workflow.

Fix this up by moving the details to the end, and adjusting the
transitions between the sections a bit.

The diff looks like a mess, but this commit is mostly just moving
things around. There are also various grammar fixes incorporated from
review.

Reviewed-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

squash! doc: application: move build/run/debug docs after overview

Commit:
doc: application: trivial whitespace fix

Fix the white space around "Naming Conventions", which is currently
appearing on its own line.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix up "running an application"

Re-work this section for clarity, making it clearer where the division
between emulated and real hardware is, and cleaning up the
instructions. Also re-work for correctness, updating Kbuild-style
flash instructions to use CMake.

Also make sure users know they can flash and run from anywhere on the
system using cmake --build.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

dbkinder commented on doc/application/application.rst:
```diff
@@ -134,19 +483,19 @@ Overview
+that are native to the host platform.
+
+CMake supports generating scripts for several build systems, but only Ninja and
+Make are tested and supported by Zephyr. After configuration, the user can
```
Previously you used the more active "you" (thank you for doing that) but now you're using the passive "the user".  It's best to be consistent, and use "you" and keep it active.  Also, we're missing the words "second stage" to go along with your intro about there being two stages.  Let's change this to:

     After configuration, you begin the second or **build** stage by executing the ...

mbolivar replied:
That makes good sense, will do.

dbkinder commented on doc/application/application.rst:
```diff
@@ -134,19 +483,19 @@ Overview
+CMake supports generating scripts for several build systems, but only Ninja and
+Make are tested and supported by Zephyr. After configuration, the user can
+begin the **build** stage by executing the generated build scripts. These build
+scripts will be able to recompile the application without involving CMake
```
change "will be able to" to "can"

mbolivar replied:
I struggle with verbosity. Thanks for helping.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+being built. Using the keyword ``PUBLIC`` would modify how other
+libraries that link with app are built. In this case, using ``PUBLIC``
+would cause libraries that link with ``app`` to also include the
+source file :file:`src/main.c`, behaviour that we surely do not want. The
```
s/behaviour/behavior/

mbolivar replied:
American it is! Fixing in a follow-on.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
-here would cause libraries that link with app to also include the
-source file src/main.c, behaviour that we surely do not want. The
-'PUBLIC' keyword could however be useful when modifying the include
+In the above code, an existing library target named ``app`` is
```
would you edit above to be ``:file:`CMakeLists.txt` ``

mbolivar replied:
No problem, will fix in a follow on.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+
+Every application must have a :file:`CMakeLists.txt` file. This
+file is the entry point, or top level, of the build
+system that builds the final :file:`zephyr.elf` image. :file:`zephyr.elf`
```
Change to:

    system.  The final :file:`zephyr.elf` image contains both the application and kernel libraries.

mbolivar replied:
I think this is a follow-on adjustment to existing documentation, and will fix it as such. Please let me know that's not correct.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+   Refer to :ref:`boards` for more information on available boards.
+
+   The Zephyr build system determines a value for :makevar:`BOARD` by checking
+   the following, in order:
```
change to (or something like this):

     in order, (when a BOARD value is found, CMake stops looking further down the list):

mbolivar replied:
Will fix and squash into existing commit

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+
+#. If your application uses a configuration file or files other than
+   the usual :file:`prj.conf` (or :file:`prj_YOUR_BOARD.conf`, where
+   ``YOUR_BOARD`` is a board name), you can now add lines setting the
```
delete "you can now"

mbolivar replied:
Verbosity again. Squashing into existing commit and thanks.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+
+   More details are available below in :ref:`application_configuration`.
+
+#. If your application has its own kernel configuration options, you can add a
```
delete "you can"

mbolivar replied:
Yup.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+   More details are available below in :ref:`application_configuration`.
+
+#. If your application has its own kernel configuration options, you can add a
+   line setting the location of the Kconfig file which defines them.
```
s/which/that/

mbolivar replied:
You're right, that's a defining clause. Thanks.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
-CMakeLists.txt file is the entry point, or toplevel of the build
-system that builds the :file:`zephyr.elf` image. :file:`zephyr.elf`
-contains both the application and the kernel libraries.
+   **This is a somewhat advanced use case; you probably don't need this.**
```
Change to:

     An (unlikely) advanced use case would be if your application has its own
     unique configuration **options** that are set differently depending on the
     build configuration.

Then continue with, "If you just want to ..."

mbolivar replied:
OK, will fix.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
+
+#. Now include the mandatory boilerplate that integrates the
+   application with the Zephyr build system on a new line, **after any
+   lines added from above steps**:
```
change to "lines added from the steps above"

mbolivar replied:
Will do.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
       project(NONE)
 
-#. Add any application source files to the 'app' target library:
+#. You can now add any application source files to the 'app' target
```
Change "You can now add" to just "Now add"

mbolivar replied:
Will do.

dbkinder commented on doc/application/application.rst:
```diff
@@ -161,53 +510,102 @@ build scripts like this:
-#. Save and close the :file:`CMakeLists.txt`.
-
-Below is an example CMakeList.txt:
+Below is a simple example CMakeList.txt:
```
change to 

     Below is a simple but complete example :file:`CMakeList.txt`:

mbolivar replied:
This also is pre-existing, so I'll fix it with an additional commit.

dbkinder commented on doc/application/application.rst:
```diff
@@ -223,105 +621,142 @@ Below is an example CMakeList.txt:
 *************************
 
-The application's kernel is configured using a set of configuration options
+The application's kernel is configured using a set of options
```
should this be "The Zephyr kernel..."

mbolivar replied:
This is a tough call in the presence of application-specific configuration options. The totality of the build configuration is no longer just Zephyr or application, but a combination of both.

I think this is an accurate change:

```
@@ -624,7 +624,7 @@ Below is a simple example :file:`CMakeList.txt`:
 Application Configuration
 *************************
 
-The application's kernel is configured using a set of options
+The application is configured using a set of options
 that can be customized for application-specific purposes.
 The Zephyr build system takes a configuration option's value from the first
 source in which it is specified.
```

dbkinder commented on doc/application/application.rst:
```diff
@@ -223,105 +621,142 @@ Below is an example CMakeList.txt:
 source in which it is specified.
 
-The available sources are (in order):
+An option's value is taken from the following available sources, in order:
```
merge this with the previous sentence:

      ... it is specifiied, taken from the following available sources, in order:

mbolivar replied:
That's better. Will do, thanks.

dbkinder commented on doc/application/application.rst:
```diff
@@ -223,105 +621,142 @@ Below is an example CMakeList.txt:
+     CONFIG_SOME_BOOL=y
 
-#. Navigate to the :file:`app`, and create the :file:`prj.conf` file. Enter:
+  To ensure that a boolean configuration option is not set, add a line
```
are you having them add a line, or are then commenting out an existing line?

mbolivar replied:
This is one of the subtleties of Kconfig.

The correct approach is to add a line which looks like a comment, but in fact ensures the option is not set.

dbkinder commented on doc/application/application.rst:
```diff
@@ -341,7 +776,8 @@ entries manually, using the configuration menu is a preferred method.
 
    * Type a numerical value in the round brackets :guilabel:`( )`.
```
change "round brackets" to "parentheses"

mbolivar replied:
Pre-existing issue, adding a follow-on commit, thanks for the catch

dbkinder reviewed (changes requested):
thanks for diving into this documentation effort!

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+   You must create a build directory yourself, and call ``cmake`` from
+   there. The Zephyr build system does not create build directories for you.
+
+The following sections describe how to create, build, and run Zephyr
```
Good fix, will do, thanks.

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+
+       $ export CONF_FILE=prj.alternate.conf
+       $ cmake -GNinja ..
+
```
You do in fact! This was a bug before this series was introduced, so I'll add a fix as a follow-on commit.

Commit:
doc: application: fix up alternate CONF_FILE

Fix the instructions for when an alternate CONF_FILE is used.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: debugging fixes

Grammatical fixes caught in review of unrelated changes.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+
+
+This section is a quick hands-on reference to start debugging your
+application with QEMU. Most content in this section is already covered on
```
Same story, this is a previously existing issue, so I'll add a fix in its own commit.

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
-The following predefined variables configure the development project:
+.. _GNU_Debugger: http://www.gnu.org/software/gdb
+
+In this quick reference you find shortcuts, specific environmental variables and
```
Previously exiting issue, will add a fix in this series

Commit:
doc: application: define ELF

The ELF format and extension is an acronym that needs a
definition. Add it.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+The simplest way to debug an application running in QEMU is using the GNU
+Debugger and setting a local GDB server in your development system through QEMU.
+
+You will need an ELF binary image for debugging purposes.  The build system
```
Sure, adding a patch to fix that.

Commit:
doc: application: fix QEMU capitalization

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+server instance. This port number can be changed for a port that best suits the
+development environment.
+
+You can run Qemu to listen for a "gdb connection" before it starts executing any
```
Adding a patch to fix pre-existing issue.

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+
+.. note::
+
+   The GDB version on the development system might not support the --tui
```
This is a pre-existing issue that I don't have a quick fix for. Since I've just moved around the problematic documentation, I don't think this needs to be addressed to merge this series.

nashif replied:
if --tui then just use GDB as is?

Commit:
doc: application: fix typo

s/remove/remote/

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+   option.
+
+If you are not using a .gdbinit file, issue the following command inside GDB to
+connect to the remove GDB server on port 1234:
```
Fixing in a follow-on

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+
+   (gdb) target remote localhost:1234
+
+Finally, The command below connects to the GDB server using the Data
```
I missed this. Thanks for catching it.

Commit:
doc: application: provide ddd install instructions

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+
+The :abbr:`DDD (Data Displayer Debugger)` may not be installed in your
+development system by default. Follow your system instructions to install
+it.
```
Adding a patch for that fix

mbolivar commented on doc/application/application.rst:
```diff
@@ -3,129 +3,478 @@
+change depending on the toolchain you are using and your cross-development
+tools.
 
 CMake
```
Good idea!

mbolivar commented on doc/application/application.rst:
```diff
@@ -134,19 +483,19 @@ Overview
 ========
 
 CMake is used to build Zephyr. A CMake build is done in two
```
Will do.

Commit:
doc: application: clarify a pronoun antecedent

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix "behavior" british spelling

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: revise app/zephyr link documentation

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: clarify KCONFIG_ROOT section

The documentation regarding application-specific Kconfig options is
unclear. Fix that.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: add rest formatting to CMakeLists.txt

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fixup for app versus zephyr config

The line between Zephyr versus application is blurry since they share
a configuration, but try to disambiguate.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: merge configuration paragraphs

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix britishism regarding parentheses

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
@dbkinder thanks for the review!

I've addressed your comments either through updating existing commits or adding on new ones where the issues were present before this series.

There's one *stet* response, to the question "are you having them add a line, or are then commenting out an existing line?"

nashif commented:
recheck
